src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationService.java

```diff
@@ -0,0 +1,224 @@
+import static org.openkilda.messaging.model.FlowDirectionType.PROTECTED_REVERSE;
+import static org.openkilda.messaging.model.FlowDirectionType.REVERSE;
+
+import org.openkilda.messaging.command.haflow.HaFlowValidationResponse;
+import org.openkilda.messaging.info.flow.FlowDumpResponse;
+import org.openkilda.messaging.info.flow.FlowValidationResponse;
+import org.openkilda.messaging.info.flow.PathDiscrepancyEntity;
+import org.openkilda.messaging.info.group.GroupDumpResponse;
+import org.openkilda.messaging.info.meter.MeterDumpResponse;
+import org.openkilda.messaging.model.FlowDirectionType;
+import org.openkilda.model.FlowPath;
+import org.openkilda.model.FlowStatus;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.HaFlowPath;
+import org.openkilda.model.PathId;
+import org.openkilda.model.SwitchId;
+import org.openkilda.model.cookie.FlowSegmentCookie;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.persistence.repositories.HaFlowRepository;
+import org.openkilda.persistence.repositories.SwitchRepository;
+import org.openkilda.rulemanager.RuleManager;
+import org.openkilda.rulemanager.SpeakerData;
+import org.openkilda.rulemanager.adapter.PersistenceDataAdapter;
+import org.openkilda.wfm.error.FlowNotFoundException;
+import org.openkilda.wfm.error.IllegalFlowStateException;
+import org.openkilda.wfm.error.SwitchNotFoundException;
+import org.openkilda.wfm.topology.flowhs.fsm.validation.SimpleSwitchRule;
+import org.openkilda.wfm.topology.flowhs.fsm.validation.SimpleSwitchRuleComparator;
+import org.openkilda.wfm.topology.flowhs.fsm.validation.SimpleSwitchRuleConverter;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.nio.file.InvalidPathException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class HaFlowValidationService {
+
+    private final HaFlowRepository haFlowRepository;
+    private final SimpleSwitchRuleConverter simpleSwitchRuleConverter = new SimpleSwitchRuleConverter();
+    private final SimpleSwitchRuleComparator simpleSwitchRuleComparator;
+    private final RuleManager ruleManager;
+    private final PersistenceManager persistenceManager;
+
+    public HaFlowValidationService(@NonNull PersistenceManager persistenceManager,
+                                   RuleManager ruleManager) {
+        this.haFlowRepository = persistenceManager.getRepositoryFactory().createHaFlowRepository();
+        this.ruleManager = ruleManager;
+        this.persistenceManager = persistenceManager;
+        SwitchRepository switchRepository = persistenceManager.getRepositoryFactory().createSwitchRepository();
+        this.simpleSwitchRuleComparator = new SimpleSwitchRuleComparator(switchRepository);
+    }
+
+    /**
+     * Check the flow status.
```

Comment:
Thanks, it is a good point but there are couple of reasons why I think the two checked exceptions were introduced here. ( btw this behaviour derived from FlowValidationFsm class..  )
We have two cases: 
1) flow not found, (we could return Optional.empty() in this case and it could be better approach instead of throwing an FlowNotFound exception.)
2) flow status is incorrect. 

So, I agree to change the logic that replaces flowNotFlound exception to the Optional.empty(); But we still need to throw an exception regarding an invalid DOWN state for the flow.

So the final code could looks like this:

`    /**
     * Finds an active {@link HaFlow} by ID.
     *
     * @param flowId the ID of the {@link HaFlow} to find
     * @return an {@link Optional} containing the active {@link HaFlow} if found, or an empty {@link Optional} otherwise
     * @throws IllegalFlowStateException if the {@link HaFlow} is found but its status is {@link FlowStatus#DOWN}
     */
     public Optional<HaFlow> findValidHaFlowById(String flowId) throws IllegalFlowStateException {
            Optional<HaFlow> flow = haFlowRepository.findById(flowId);
            if (flow.isPresent() && FlowStatus.DOWN.equals(flow.get().getStatus())) {
                throw new IllegalFlowStateException(flowId);
            }
            return flow;
        }`

Reply:
I just asked, I don't insist on Optional. Sometimes it just not convenient on the API user's side to wrap with try-catch, so people just propagate the exception further and don't really use the exception type (maybe the message is still important), catch and rethrow runtime, or other stuff. If it is needed to differentiate types, then all is good.
Thanks for the explanation.